system/lib/pthread/library_pthread.c

```diff
@@ -457,11 +540,73 @@ void EMSCRIPTEN_KEEPALIVE emscripten_main_thread_process_queued_calls()
 	// This must be before pthread_mutex_lock(), since pthread_mutex_lock() can call back to this function.
 	bool_inside_nested_process_queued_calls = 1;
 	pthread_mutex_lock(&call_queue_lock);
-	for (int i = 0; i < call_queue_length; ++i)
-		_do_call(call_queue[i]);
-	call_queue_length = 0;
-	bool_inside_nested_process_queued_calls = 0;
+	int head = emscripten_atomic_load_u32((void*)&call_queue_head);
+	int tail = emscripten_atomic_load_u32((void*)&call_queue_tail);
+	while (head != tail)
+	{
+		// Assume that the call is heavy, so unlock access to the call queue while it is being performed.
```

Comment:
No, we don't want to re-enter here under any circumstances, that might invert the order of execution of proxied calls. If executing one call we happened to re-enter and start executing another (when the first one is still on the stack), the second one would then finish before the first one. The queue order must be preserved in all cases.